lib/mongoid/matcher/type.rb

```diff
@@ -0,0 +1,86 @@
+module Mongoid
+  module Matcher
+
+    # @see https://docs.mongodb.com/manual/reference/operator/query/type/
+    #
+    # @api private
+    module Type
+      module_function def matches?(exists, value, condition)
+        p value
+        case condition
+        when Array
+          condition.any? do |v|
+            condition = v
+          end
+        when 1
+          # Double
+          Float === value
+        when 2
+          # String
+          String === value
+        when 3
+          # Object
+          Object === value
+        when 4
+          # Array
+          Array === value
+        when 5
+          # Binary data
+          Binary === value
+        when 6
+          # Undefined
+          BSON::Undefined === value
+        when 7
+          # ObjectId
+          BSON::ObjectId === value
+        when 8
+          # Boolean
+          !!value === value
+        when 9
+          # Date
+          Date === value
```

Comment:
This is most likely supposed to be a timestamp type, therefore  I expect `Time === value` to be more correct and potentially `Time === value || DateTime === value || Date === value` could be appropriate also.

Reply:
There is a Timestamp check later on, so I'll add that logic there.